JetStreamCompose/jetstream/src/main/java/com/google/jetstream/data/repositories/MovieRepositoryImpl.kt

```diff
@@ -33,132 +33,167 @@ import com.google.jetstream.data.util.StringConstants.Movie.Reviewer.DefaultCoun
 import com.google.jetstream.data.util.StringConstants.Movie.Reviewer.DefaultRating
 import com.google.jetstream.data.util.StringConstants.Movie.Reviewer.FreshTomatoes
 import com.google.jetstream.data.util.StringConstants.Movie.Reviewer.ReviewerName
+import kotlinx.coroutines.Dispatchers
 import kotlinx.coroutines.flow.Flow
 import kotlinx.coroutines.flow.flow
-import kotlinx.coroutines.flow.flowOf
+import kotlinx.coroutines.flow.map
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+import kotlinx.coroutines.withContext
 import kotlinx.serialization.json.Json
 import javax.inject.Inject
 
-class MovieRepositoryImpl @Inject constructor(assetsReader: AssetsReader) : MovieRepository {
+class MovieRepositoryImpl @Inject constructor(
+    private val assetsReader: AssetsReader
+) : MovieRepository {
 
-    private val mostPopularMovies: List<MoviesResponseItem> =
-        assetsReader.getJsonDataFromAsset(StringConstants.Assets.MostPopularMovies).map {
-            Json.decodeFromString<List<MoviesResponseItem>>(it)
-        }.getOrDefault(emptyList())
+    private val mostPopularMovies = flow {
+        val list = withContext(Dispatchers.IO) {
+            assetsReader.getJsonDataFromAsset(StringConstants.Assets.MostPopularMovies).map {
+                Json.decodeFromString<List<MoviesResponseItem>>(it)
+            }.getOrDefault(emptyList())
+        }
+        emit(list)
+    }
 
+    private val inTheatersMovies = flow {
+        val list = withContext(Dispatchers.IO) {
```

Comment:
Sharing these `flow`s among other functions won't create any cache. Each `flow` will subscribe to it's original source, which means it will load the data from the assets **multiple times**.

Reply:
Thanks. I misunderstood. I introduced a cache layer to prevent flow's subscriptions load assets multiple times.